distribution/lib/Standard/Database/0.0.0-dev/src/Internal/JDBC_Connection.enso

```
@@ -120,31 +118,61 @@ type JDBC_Connection
-                Managed_Resource.bracket (java_connection.prepareStatement insert_template) .close stmt->
-                    num_rows = table.row_count
-                    columns = table.columns
-                    check_rows updates_array expected_size =
-                        updates = Vector.from_polyglot_array updates_array
-                        if updates.length != expected_size then Panic.throw <| Illegal_State.Error "The batch update unexpectedly affected "+updates.length.to_text+" rows instead of "+expected_size.to_text+"." else
-                            updates.each affected_rows->
-                                if affected_rows != 1 then
-                                    Panic.throw <| Illegal_State.Error "A single update within the batch unexpectedly affected "+affected_rows.to_text+" rows."
-                    0.up_to num_rows . each row_id->
-                        values = columns.map col-> col.at row_id
-                        set_statement_values stmt statement_setter values
-                        stmt.addBatch
-                        if (row_id+1 % batch_size) == 0 then check_rows stmt.executeBatch batch_size
-                    if (num_rows % batch_size) != 0 then check_rows stmt.executeBatch (num_rows % batch_size)
+            Managed_Resource.bracket (java_connection.setAutoCommit False) (_ -> java_connection.setAutoCommit default_autocommit) _->
+                action
+
+    ## PRIVATE
+       Performs the given action within a transaction.
+       Once the action is completed, the transaction is committed.
+       If a panic escapes from the action, the transaction is rolled-back and
+       closed.
+    run_within_transaction : Any -> Any
+    run_within_transaction self ~action =
+        self.run_without_autocommit <|
+            self.with_connection java_connection->
+                handle_panic caught_panic =
+                    java_connection.rollback
```

What happens if the rollback throws another Panic?

---

I suppose the rollback panic will take precedence. I'm not sure which one it should be here, but it feels like silencing the rollback panic is wrong, so I think I'd keep it that way. I'll add a note to the docs.

---

Would it be possible to catch it, and combine the messages into one?

---

No, because the errors do not form a monoid - we don't know what the panic returned from within the transaction may be - it can be of any type.

We could probably rely on `Throwable::addSuppressed` to indicate that one exception was suppressed by another, this seems to be preferred way to do this in Java, but Enso currently does not have much support for this. I think we can add it if necessary though. Shall we create a ticket to tackle this? Enso did not really have a kind of 'finally' clause so it was not that useful - now I also added `Panic.with_finalizer` where this could be useful as well. But it doesn't feel like it's a highest priority thing, so I'd be hesitant to spend too much time on it before Beta Release.

---

Yes, agreed, we should have a ticket for it. It's useful in general (and in this case, when you want to know if the rollback failed), but doesn't seem urgent since it's very unlikely to fail in practice.

---

Added a ticket #6440